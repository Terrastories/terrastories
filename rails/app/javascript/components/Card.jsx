import React, { Component } from "react";
import PropTypes from 'prop-types';
import StoryList from "./StoryList";

class Card extends Component {

  constructor(props) {
    super(props);
    this.state = {
      isToggleOn: true
    };
  }

  static propTypes = {
    filterMap: PropTypes.object,
    categories: PropTypes.array,
    clearFilteredStories: PropTypes.func,
    handleFilter: PropTypes.func,
    user: PropTypes.object,
    stories: PropTypes.array,
    handleStoriesChanged: PropTypes.func,
    onStoryClick: PropTypes.func,
    logo_path: PropTypes.string,
    activeStory: PropTypes.object,
    filterCategory: PropTypes.string,
    filterItem: PropTypes.string,
    handleFilterCategoryChange: PropTypes.func,
    handleFilterItemChange: PropTypes.func,
    itemOptions: PropTypes.array
  };

  static defaultProps = {
    filterMap: {},
    categories: [],
    clearFilteredStories: () => { },
    handleFilter: () => { }
  }

  handleTray = () => {
    this.setState(prevState => ({
      isToggleOn: !prevState.isToggleOn
    }));
  }

  renderUserInformation = () => {
    if (this.props.user && this.props.user.role === 'editor') {
      return (
        <ul>
          <li>{I18n.t("hello")} {this.props.user.email} (<a href={`/${I18n.currentLocale()}`}>{I18n.t("back_to_welcome")}</a>)</li>
          <li><a href={`/admin?locale=${I18n.currentLocale()}`}>{I18n.t("admin_page")}</a></li>
        </ul>
      );
    } else if (this.props.user) {
      return (
        <ul>
          <li>{I18n.t("hello")} {this.props.user.email} (<a href={`/${I18n.currentLocale()}`}>{I18n.t("back_to_welcome")}</a>)</li>
        </ul>
      );
    } else {
      return (
        <ul>
          <li><a href={`/${I18n.currentLocale()}`}>{I18n.t("back_to_welcome")}</a></li>
        </ul>
      );
    }
  }

  render() {
    return (
      <div className={this.state.isToggleOn ? 'cardContainer onCanvas' : 'cardContainer offCanvas'}>
        <div className="tab" onClick={this.handleTray}>
          <div className="arrow" />
        </div>
        <div className="closeMe" onClick={this.handleTray} />
        <div className="card">
          <div className="bar">
            <div className="card--logo">
              <img src={this.props.logo_path} alt="Terrastories" />
            </div>

            <StoryList
              activeStory={this.props.activeStory}
              stories={this.props.stories}
              handleStoriesChanged={this.props.handleStoriesChanged}
              onStoryClick={this.props.onStoryClick}
              handleFilter={this.props.handleFilter}
              clearFilteredStories={this.props.clearFilteredStories}
              filterMap={this.props.filterMap}
              categories={this.props.categories}
              filterCategory={this.props.filterCategory}
              filterItem={this.props.filterItem}
              handleFilterCategoryChange={this.props.handleFilterCategoryChange}
              handleFilterItemChange={this.props.handleFilterItemChange}
              itemOptions={this.props.itemOptions}
            />

            <div className="card--tasks">
              {this.renderUserInformation()}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default Card;
