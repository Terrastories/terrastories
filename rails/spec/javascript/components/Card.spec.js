import React from 'react';
import { shallow } from 'enzyme';
import Card from '../../../app/javascript/components/Card';

describe('Card component', () => {
  beforeEach(() => {
    global.I18n = {
      currentLocale: jest.fn(),
      t: jest.fn(),
    };
  })

  it('Displays correctly', () => {
    const wrapper = shallow(
      <Card />,
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('StoryList')).toHaveLength(1)
  })

  describe('Given a user props', () => {
    describe('With an editor role', () => {
      it('Renders the correct information', () => {
        const wrapper = shallow(
          <Card user={{ role: 'editor' }} />,
        );

        expect(wrapper.find('.card--tasks').childAt(0)).toHaveLength(1);
        expect(global.I18n.currentLocale).toHaveBeenCalledTimes(2);
        expect(global.I18n.t).toHaveBeenCalledWith('hello');
        expect(global.I18n.t).toHaveBeenCalledWith('back_to_welcome');
        expect(global.I18n.t).toHaveBeenCalledWith('admin_page');
      });
    })

    describe('With no role', () => {
      it('Renders the correct information', () => {
        const wrapper = shallow(
          <Card user={{ role: null }} />,
        );

        expect(wrapper.find('.card--tasks').childAt(0)).toHaveLength(1);
        expect(global.I18n.currentLocale).toHaveBeenCalledTimes(1);
        expect(global.I18n.t).toHaveBeenCalledWith('hello');
        expect(global.I18n.t).toHaveBeenCalledWith('back_to_welcome');
        expect(global.I18n.t).not.toHaveBeenCalledWith('admin_page');
      });
    })

    describe('That is null/empty', () => {
      it('Renders the correct information', () => {
        const wrapper = shallow(
          <Card />,
        );

        expect(wrapper.find('.card--tasks').childAt(0)).toHaveLength(1);
        expect(global.I18n.currentLocale).toHaveBeenCalledTimes(1);
        expect(global.I18n.t).not.toHaveBeenCalledWith('hello');
        expect(global.I18n.t).toHaveBeenCalledWith('back_to_welcome');
        expect(global.I18n.t).not.toHaveBeenCalledWith('admin_page');
      });
    })
  });

  it('Toggles the onCanvas/offCanvas CSS classes', () => {
    const wrapper = shallow(
      <Card />,
    );

    expect(wrapper.find('.cardContainer').hasClass('onCanvas')).toBe(true);
    expect(wrapper.find('.cardContainer').hasClass('offCanvas')).toBe(false);
    wrapper.find('.tab').simulate('click');
    expect(wrapper.find('.cardContainer').hasClass('onCanvas')).toBe(false);
    expect(wrapper.find('.cardContainer').hasClass('offCanvas')).toBe(true);
    wrapper.find('.tab').simulate('click');
    expect(wrapper.find('.cardContainer').hasClass('onCanvas')).toBe(true);
    expect(wrapper.find('.cardContainer').hasClass('offCanvas')).toBe(false);
  })
});
