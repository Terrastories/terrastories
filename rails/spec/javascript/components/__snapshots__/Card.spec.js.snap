// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Displays correctly 1`] = `
<div
  className="cardContainer onCanvas"
>
  <div
    className="tab"
    onClick={[Function]}
  >
    <div
      className="arrow"
    />
  </div>
  <div
    className="closeMe"
    onClick={[Function]}
  />
  <div
    className="card"
  >
    <div
      className="bar"
    >
      <div
        className="card--logo"
      >
        <img
          alt="Terrastories"
        />
      </div>
      <StoryList
        categories={Array []}
        clearFilteredStories={[Function]}
        filterMap={Object {}}
        handleFilter={[Function]}
      />
      <div
        className="card--tasks"
      >
        <ul>
          <li>
            <a
              href="/undefined"
            />
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
